File: DataSets/DataSet-New/LeetCodeTasks - Develop/LeetCodeTasks/3123_find-edges-in-shortest-paths/solutions/java/solution.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.PriorityQueue;

class Solution {
    public boolean[] findAnswer(int n, int[][] edges) {
        // Initialize graph
        List<List<int[]>> graph = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            graph.add(new ArrayList<>());
        }
        for (int[] edge : edges) {
            int u = edge[0], v = edge[1], w = edge[2];
            graph.get(u).add(new int[]{v, w});
            graph.get(v).add(new int[]{u, w});
        }

        // Distance from the source (node 0)
        long[] dist = new long[n];
        Arrays.fill(dist, Long.MAX_VALUE);
        dist[0] = 0;
        
        // Dijkstra's algorithm to find shortest paths from node 0
        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> Long.compare(a[1], b[1]));
        pq.offer(new int[]{0, 0}); // (node, dist[node])

        while (!pq.isEmpty()) {
            int[] current = pq.poll();
            int node = current[0];
            long d = current[1];
            
            if (d > dist[node]) continue; // Ignore outdated distance

            for (int[] edge : graph.get(node)) {
                int neighbor = edge[0];
                long weight = edge[1];
                if (dist[node] + weight < dist[neighbor]) {
                    dist[neighbor] = dist[node] + weight;
                    pq.offer(new int[]{neighbor, (int)dist[neighbor]});
                }
            }
        }

        // Check which edges are on any of the shortest paths
        boolean[] answer = new boolean[edges.length];
        for (int i = 0; i < edges.length; i++) {
            int u = edges[i][0], v = edges[i][1], w = edges[i][2];
            if (dist[u] + w == dist[v] || dist[v] + w == dist[u]) {
                answer[i] = true;
            }
        }
        
        return answer;
    }
}

File: DataSets/DataSet-New/LeetCodeTasks - Develop/LeetCodeTasks/3123_find-edges-in-shortest-paths/solutions/python3/solution.py
from heapq import heappop, heappush
from collections import defaultdict, deque

class Solution:
    def findAnswer(self, n: int, edges: List[List[int]]) -> List[bool]:
        # Create adjacency list for graph representation
        graph = defaultdict(list)
        for index, (u, v, w) in enumerate(edges):
            graph[u].append((v, w, index))
            graph[v].append((u, w, index))

        # Dijkstra's algorithm to find shortest paths from source node 0
        def dijkstra(source, n):
            dist = [float('inf')] * n
            dist[source] = 0
            min_heap = [(0, source)]
            while min_heap:
                d, u = heappop(min_heap)
                if d > dist[u]:
                    continue
                for v, weight, idx in graph[u]:
                    if dist[u] + weight < dist[v]:
                        dist[v] = dist[u] + weight
                        heappush(min_heap, (dist[v], v))
            return dist

        # Find the shortest path distance from node 0 to all nodes
        distances = dijkstra(0, n)

        # Find the shortest path distance from node n-1 to all nodes (reversed Dijkstra)
        reverse_distances = dijkstra(n-1, n)

        # Total distance from 0 to n-1 is the shortest path distance
        shortest_path_distance = distances[n-1]

        # Determine which edges are on the shortest path
        answer = [False] * len(edges)
        for index, (u, v, weight) in enumerate(edges):
            if distances[u] + weight + reverse_distances[v] == shortest_path_distance:
                answer[index] = True
            elif distances[v] + weight + reverse_distances[u] == shortest_path_distance:
                answer[index] = True

        return answer
